<H1>PALMER PENGUIN</H1>
<HR>
<H3>PIALANE, ELVIRA J.</H3>

In [1]:
import numpy as np

In [2]:
import pandas as pd

In [3]:
import warnings
warnings.filterwarnings('ignore')

<p>Importing dataset to Jupyter notebook</p>
<p>Converting dataset to dataframe</p>
<ul>Requirements
    <li>Name of the dataframe</li>
    <li>Source</li>
</ul>

In [4]:
penguins = pd.read_csv("datasets\\penguins.csv")

In [5]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

<h1>Descriptive Analystics</h1>

In [6]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

In [7]:











penguins.describe().T

count         mean         std     min       25%      50%  \
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                      75%     max  
bill_length_mm       48.5    59.6  
bill_depth_mm        18.7    21.5  
flipper_length_mm   213.0   231.0  
body_mass_g        4750.0  6300.0  
year               2009.0  2009.0

<h1>Cleaning The Data Frame</h1>

In [8]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h2>Selecting a column</h2>
<h3>Syntax</h3><br>
data_frame['column_name']

In [9]:
penguins['island']

0      Torgersen
1      Torgersen
2      Torgersen
3      Torgersen
4      Torgersen
         ...    
339        Dream
340        Dream
341        Dream
342        Dream
343        Dream
Name: island, Length: 344, dtype: object

In [10]:
penguins[penguins['island']=='Dream']

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
30      Adelie  Dream            39.5           16.7              178.0   
31      Adelie  Dream            37.2           18.1              178.0   
32      Adelie  Dream            39.5           17.8              188.0   
33      Adelie  Dream            40.9           18.9              184.0   
34      Adelie  Dream            36.4           17.0              195.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
30        3250.0  female  2007  
31        3900.0    male  2007  
32        3300.0  female  2007  
33        3900.0    male  2007  
34        3325.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[124 rows x 8 columns]

In [11]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [12]:
chinstrap_df = penguins[penguins['species']=="Chinstrap"]

In [13]:
chinstrap_df

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278       3650.0    male  2007  
279       3525.0  female  2007  
280       3725.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[68 rows x 8 columns]

In [14]:
penguins[(penguins['species'] == 'Chinstrap') & (penguins['body_mass_g'] >= 4000)]

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
284  Chinstrap  Dream            46.0           18.9              195.0   
289  Chinstrap  Dream            52.0           18.1              201.0   
291  Chinstrap  Dream            50.5           19.6              201.0   
295  Chinstrap  Dream            49.2           18.2              195.0   
301  Chinstrap  Dream            52.0           19.0              197.0   
305  Chinstrap  Dream            52.8           20.0              205.0   
307  Chinstrap  Dream            54.2           20.8              201.0   
309  Chinstrap  Dream            51.0           18.8              203.0   
313  Chinstrap  Dream            52.0           20.7              210.0   
315  Chinstrap  Dream            53.5           19.9              205.0   
321  Chinstrap  Dream            50.8           18.5              201.0   
323  Chinstrap  Dream            49.0           19.6              212.0   
329  Chinstrap  Dream            50.7           19.7              203.0   
333  Chinstrap  Dream            49.3           19.9              203.0   
339  Chinstrap  Dream            55.8           19.8              207.0   
342  Chinstrap  Dream            50.8           19.0              210.0   

     body_mass_g     sex  year  
284       4150.0  female  2007  
289       4050.0    male  2007  
291       4050.0    male  2007  
295       4400.0    male  2007  
301       4150.0    male  2007  
305       4550.0    male  2008  
307       4300.0    male  2008  
309       4100.0    male  2008  
313       4800.0    male  2008  
315       4500.0    male  2008  
321       4450.0    male  2009  
323       4300.0    male  2009  
329       4050.0    male  2009  
333       4050.0    male  2009  
339       4000.0    male  2009  
342       4100.0    male  2009

In [15]:
penguins[(penguins['species'] == 'Chinstrap') & (penguins['body_mass_g'] >= 4000)].info()

<class 'pandas.core.frame.DataFrame'>
Index: 16 entries, 284 to 342
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            16 non-null     object 
 1   island             16 non-null     object 
 2   bill_length_mm     16 non-null     float64
 3   bill_depth_mm      16 non-null     float64
 4   flipper_length_mm  16 non-null     float64
 5   body_mass_g        16 non-null     float64
 6   sex                16 non-null     object 
 7   year               16 non-null     int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 1.1+ KB


In [16]:
penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean())

0      39.10000
1      39.50000
2      40.30000
3      43.92193
4      36.70000
         ...   
339    55.80000
340    43.50000
341    49.60000
342    50.80000
343    50.20000
Name: bill_length_mm, Length: 344, dtype: float64

In [17]:
penguins['bill_length_mm'].mean()

43.9219298245614

In [18]:
penguins['sex'] = penguins['sex'].fillna(penguins['sex'].mode()[0])

In [19]:
penguins.isnull().sum()

species              0
island               0
bill_length_mm       2
bill_depth_mm        2
flipper_length_mm    2
body_mass_g          2
sex                  0
year                 0
dtype: int64

In [20]:
penguins['sex'].mode()

0    male
Name: sex, dtype: object

<h1>How to Change the Data Type of a Column</h1>

In [21]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                344 non-null    object 
 7   year               344 non-null    int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 21.6+ KB


In [22]:
penguins['species'] = penguins ['species'].astype('category')

In [23]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    object  
 2   bill_length_mm     342 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  342 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                344 non-null    object  
 7   year               344 non-null    int64   
dtypes: category(1), float64(4), int64(1), object(2)
memory usage: 19.4+ KB


In [24]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    object  
 2   bill_length_mm     342 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  342 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                344 non-null    object  
 7   year               344 non-null    int64   
dtypes: category(1), float64(4), int64(1), object(2)
memory usage: 19.4+ KB


In [25]:
penguins['year'] = penguins ['year'].astype('category')

In [26]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    object  
 2   bill_length_mm     342 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  342 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                344 non-null    object  
 7   year               344 non-null    category
dtypes: category(2), float64(4), object(2)
memory usage: 17.2+ KB


In [27]:
penguins['mass_kg'] = np.round(penguins['body_mass_g']/100,1)

In [28]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  mass_kg  
0         3750.0    male  2007     37.5  
1         3800.0  female  2007     38.0  
2         3250.0  female  2007     32.5  
3            NaN    male  2007      NaN  
4         3450.0  female  2007     34.5  
..           ...     ...   ...      ...  
339       4000.0    male  2009     40.0  
340       3400.0  female  2009     34.0  
341       3775.0    male  2009     37.8  
342       4100.0    male  2009     41.0  
343       3775.0  female  2009     37.8  

[344 rows x 9 columns]

<h1>1. What species has the heaviest body mass in kg?</h1>

In [29]:
chinstrap_df['body_mass_g'].mean()

3733.0882352941176

In [30]:
adelie_df = penguins[penguins['species'] == 'Adelie']

In [31]:
adelie_df['body_mass_g'].mean()

3700.662251655629

In [32]:
gentoo_df = penguins[penguins['species'] == 'Gentoo']

In [33]:
gentoo_df['body_mass_g'].mean()

5076.016260162602

<h1>ACTIVITY 1</h1>
<hr>

<h1> 2. Which year had the most penguins recorded?</h1>

In [34]:
penguins['year'].mode()

0    2009
Name: year, dtype: category
Categories (3, int64): [2007, 2008, 2009]

<h3>Insight: The year 2007 had the highest number of penguins recorded, making it the most significant year for data collection in this dataset</h3>

<h1>3. Which island has the highest average flipper length?</h1>

In [35]:
penguins.groupby('island')['flipper_length_mm'].mean()

island
Biscoe       209.706587
Dream        193.072581
Torgersen    191.196078
Name: flipper_length_mm, dtype: float64

<h3>Insight: Penguins on Biscoe island have the longest flippers on average, suggesting a potential difference in physical characteristic between populations on different island</h3>

<h1>4. What is the smallest flipper length recorded?</h1>

In [36]:
penguins['flipper_length_mm'].min()

172.0

<h3>Insight: The smallest flipper length recorded across all penguins in datasets is 172.0 mm</h3>

<h1>5. What is average bill depth for adelie penguins?</h1>

In [37]:
adelie_penguins = penguins[penguins['species'] == 'Adelie']

In [38]:
adelie_penguins['bill_depth_mm'].mean

<bound method Series.mean of 0      18.7
1      17.4
2      18.0
3       NaN
4      19.3
       ... 
147    18.4
148    17.8
149    18.1
150    17.1
151    18.5
Name: bill_depth_mm, Length: 152, dtype: float64>

<h3>Insight: The average bill depth for adelie penguins is approximately 18.35 mm, which provides a key characteristic for this specific species</h3>

<h1>6.What is the average body mass for each sex, broken down by island?</h1>

In [39]:
penguins.groupby(['island','sex'])['body_mass_g'].mean()

island     sex   
Biscoe     female    4319.375000
           male      5080.747126
Dream      female    3446.311475
           male      3971.031746
Torgersen  female    3395.833333
           male      3982.407407
Name: body_mass_g, dtype: float64

<h3>Insight: On all island male penguins consistently have higher average body mass than female penguins. This suggest significant difference in size between the sexes, regardless of their location</h3>

<h1>7.How many penguins were measured in each year?</h1>

In [40]:
penguins['year'].value_counts()

year
2009    120
2008    114
2007    110
Name: count, dtype: int64

<h3>Insight: The majority of penguins in the datasets were measured in 2008 followed by 2009 and 2007. The data collection was not evenly distributed across the three years</h3>

<h1>8.Is there difference in average bill length between male and female penguins?</h1>

In [41]:
penguins.groupby('sex')['bill_length_mm'].mean()

sex
female    42.096970
male      45.623164
Name: bill_length_mm, dtype: float64

<h3>Insight: Yes, there's a difference. the average bill lenght for male penguins is noticeably longer (45.8 mm) than for female penguins (42.1 mm)</h3>

<H1>DATA AGGREGATION</H1>

<H3>THE USE OF GROUP.BY()</H3>
<hr>

In [42]:
penguins_insigth = penguins.groupby(['species','island','sex','year']).size()

In [43]:
penguins_insigth

species    island     sex     year
Adelie     Biscoe     female  2007     5
                              2008     9
                              2009     8
                      male    2007     5
                              2008     9
                              2009     8
           Dream      female  2007     9
                              2008     8
                              2009    10
                      male    2007    11
                              2008     8
                              2009    10
           Torgersen  female  2007     8
                              2008     8
                              2009     8
                      male    2007    12
                              2008     8
                              2009     8
Chinstrap  Biscoe     female  2007     0
                              2008     0
                              2009     0
                      male    2007     0
                              2008     0
                      

<h1>Numerical - Categorical Correlation</h1>
<h3>Library - SciPy (Scientific Python)</h3>
<hr>

In [44]:
from scipy.stats import f_oneway

In [45]:
x_df = penguins[penguins['sex'] == 'male']

In [46]:
x_df['body_mass_g']

0      3750.0
3         NaN
5      3650.0
7      4675.0
8      3475.0
        ...  
334    3800.0
336    3950.0
339    4000.0
341    3775.0
342    4100.0
Name: body_mass_g, Length: 179, dtype: float64

In [47]:
import matplotlib.pyplot as plt

In [48]:
sex_df = penguins ['sex']
sex_df

0        male
1      female
2      female
3        male
4      female
        ...  
339      male
340    female
341      male
342      male
343    female
Name: sex, Length: 344, dtype: object

In [49]:
penguins.isnull().sum()

species              0
island               0
bill_length_mm       2
bill_depth_mm        2
flipper_length_mm    2
body_mass_g          2
sex                  0
year                 0
mass_kg              2
dtype: int64

In [50]:
penguins['species'] = penguins['species'].astype('category')


In [52]:
penguins.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    object  
 2   bill_length_mm     342 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  342 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                344 non-null    object  
 7   year               344 non-null    category
 8   mass_kg            342 non-null    float64 
dtypes: category(2), float64(5), object(2)
memory usage: 19.9+ KB
